mesa: enable freedreno virtio kernel mode driver

kuruczgy opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This change enables a NixOS guest to utilize a virtio-gpu device exposing a drm native context capset. Host qemu still needs some extra patches to get this to work, but as far as the guest goes this is the only change needed.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

K900 commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -92,6 +92,11 @@
+
+# This is ignored when freedreno is not being built. (e.g. on x86_64)
+# virtio is needed for a guest to use DRM native context.
+, freedrenoKernelModeDrivers ? ["msm" "virtio"]
```
Does this need to be an option?

kuruczgy replied:
The full list of available options is `'msm', 'kgsl', 'virtio', 'wsl'`, though not sure if `kgsl` or `wsl` are actually implemented or useful. Did not want to make it difficult for someone adventurous to enable those if they want to experiment.

But we could just hardcode `(lib.mesonOption "freedreno-kmds" "msm,virtio")` if you prefer that, I don't have a problem with that.

K900 replied:
What's the cost to just build all of them?

kuruczgy replied:
I guess nothing other than a bit of extra code size. Pushed an update to enable them all unconditionally.

K900 replied:
I guess next question then: will this break with Mesa 24.3?

kuruczgy replied:
No idea, tried to ask in #freedreno on IRC but did not get a response to this particular question. I personally would be surprised though if they broke this any time soon, it was at least confirmed to me that all 4 options have some use-case (though I was told that WSL requires some extra tooling to work beside just enabling this), so even if they ever remove one of them I expect it to come with an ample deprecation warning.

K900 replied:
Next time, please just check the build files. https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/meson_options.txt?ref_type=heads#L222

kuruczgy committed:
mesa: enable all freedreno kernel mode driver

`virtio` is used with a virtio-gpu device supporting drm native context.
I don't know the use-cases for `wsl` and `kgsl`, but let's just enable
them too.

kuruczgy commented:
I see that `ofborg-eval` failed, though the error does not seem related to this PR. (But let me know if you think it is.)
